src/dusk.rs

```diff
@@ -213,6 +213,44 @@ impl JubJubExtended {
         let p = JubJubAffine::from(self);
         [p.u, p.v]
     }
+
+    /// Hash an arbitrary slice of bytes to a point on the elliptic curve and
+    /// in the prime order subgroup.
+    ///
+    /// This algorithm uses rejection sampling to hash to a point on the curve:
+    /// The input together with a counter are hashed into an array of 32 bytes.
+    /// If the hash is a canonical representation of a point on the curve and
+    /// a member of the prime-order subgroup, we return it. If not, we increment
+    /// the counter and try to de-serialize again.
+    ///
+    /// **Note:** This implementation of `hash_to_point` is not ideal, in the
+    /// long run we want to implement an algorithm outlined
+    /// [here](https://datatracker.ietf.org/doc/html/rfc9380), but we start with
+    /// this implementation in order to be able to use the API already.
+    pub fn hash_to_point(input: &[u8]) -> Self {
```

Comment:
A point must always exist for a generator, otherwise it is unusable in terms of protocol & circuit design.

With that said, I think it's best we generate a jubjub scalar, take a generator as argument, and multiply the two.

Reply:
What exactly does that approach solve? We could still hit the identity point and additionally we introduce a relation to the generator point given as an argument and the point that we output, which can be used as a attack vector

Reply:
This is not related to the identity point shield; it is to the possible applications of a hash to curve implementation (i.e. it must exist for a generator)

Reply:
I'm not sure I understand what you are getting at. Why does this need to exist for a generator?
As an example of a use-case: this `hash_to_point` will used to generate a new generator point for eash user in the zedger protocol. This is because each zedger secret key consist of a unique secret scalar and a unique secret generator point.

Reply:
Hmm that's interesting. The generator will be secret too? I wonder what ECC operations they can perform. Can you link as comment on the code that we expect to have secret generator for zedger, and some specs of zedger describing the usage of this point?

Even if the scalar & generator are secret for the protocol, they might have to be known by the owner of the secret. Ain't that the case? Otherwise, it seems pointless to me to have an ECC point

Reply:
The secret key (scalar and point) is known to the user, but the scalar and point will be generated by a different source of randomness (the scalar by a seed and the point by a randomly sampled element `z`). See https://github.com/dusk-network/zedger-specs/blob/main/specification/build/main.pdf section 6.4.1 holder secret key.

Reply:
IIUC the holder needs to know `G`. If we generate the point randomly, it's impossible to infer `G`. Did I miss something? `6.4.1.Holder.7` states that the holder must store his `G`

Reply:
The nothing up my sleeve algorithm is exactly the same algorithm that I use in this PR. And creating generators is tricky exactly because there shouldn't be any relations to existing generator points. This is the reason why I think we should not use a known generator point in `hash_to_point`.

Reply:
We don't have a canonical algorithm to generate safe numbers. For this [G' here](https://github.com/dusk-network/jubjub/blob/1f832dc3839e2c7d1497abac0327674747e0fd2b/src/dusk.rs#L58-L60) we used a counter of `18`. In this case, we are incrementing the counter until we find a valid point (with high confidence it is going to be `0`, most of the times).

My input on this PR (that naturally can be bypassed) is to run the specs discussion with the research team, document the decided algorithm, and only then go for the implementation (with a link to the docs agreed on the aforementioned step)

Reply:
Also, as I specify in the Note section, we will still update the algorithm with the one used in https://datatracker.ietf.org/doc/html/rfc9380 which was suggested by Dmitry as the ideal approach.

As a 'naive' approach Dmitry suggested the following:
- hash to a scalar
- use that as a x-coordinate
- generate a y-coordinate according to the jubjub curve equation
- and check if this point is part of the subgroup with prime-order `r`
- if the point is not part of the subgroup, increment x + 1 and try again.

This algorithm is not suitable in my eyes though, since there are two possible y-coordinates for each x.

This is why I opted for the nothing-up-my-sleeve algorithm that we already use for `GENERATOR_NUMS` as a place-holder until we have the time to implement the ideal approach.

I asked Dmitry if the modified version of his naive approach is alright, I can tag you in the conversation.

Reply:
The nothing up my sleeve algorithm doesn't fix a counter (18 is just the one that happens to fit for this input). The whole point of nothing up my sleeve is that the relation from one generator to the other is unknown.

Reply:
I'll go ahead and merge this PR so that the zedger implementation gets unblocked.